### ISSUE TITLE
Decouple from pretender, support MSW others?

### ISSUE THREAD

**Issue by andrewluetgers** [miragejs/miragejs#1013]
Hi I am coming to this project from using MSW and realized organically that mocking the api is the best solution vs one off snapshots. This is clearly the project for doing such things, very nice! I however really appreciate how MSW works and has integrated nicely with my testing and Storybook development workflows. How hard would it be for me to get this working in MSW instead of leaning on pretender? Any guidance or previous attempts at this would be greatly appreciated. 

I think going forward this project would grow even faster if this was configurable, mock with pretender, or MSW or whatever. Case in point, Kent C. Dodds seems to let the issue of how, (technically) you mock requests, blind him to the benefits of why domain modeling and mocking an api with mirage.js is superior to other methods. https://kentcdodds.com/calls/01/13/mirage-js-over-msw Making this issue go away could mean many more people using mirage.js

**Comment by IanVS** [comment]
I personally would be very on board with this, but I have no idea how tightly coupled mirage is to pretender.  From what I can tell, @samselikoff wouldn't be opposed to it, and may even have some thoughts on how it could be done.  Sam, a bit of guidance here could go a long way towards moving this project forward, IMHO.

**Comment by jneurock** [comment]
I'm not aware of previous attempts but I have thought about the same thing and would love to figure out how much effort it would be. Maybe something nice to work on over the holiday break.

**Comment by samselikoff** [comment]
I haven't been personally maintaining Mirage for a while now but happy to give any advice. Maybe a call in the future would be helpful if folks were really willing to take on the work.

Here's some history:

- https://github.com/miragejs/miragejs/issues/210
- https://github.com/miragejs/miragejs/pull/699/files
- https://github.com/miragejs/miragejs/pull/320

I'll have to ask @ryanto if there was more code on our machines somewhere but the general idea was to make an async  `server.handle(request)` method that would return the response as just a JS response object (req/res signature undecided, ideally something based on browser APIs or copy from express). Then use something of an adapter pattern so people can hook that into pretender (backwards-compat), or use it within an express route handler (to run it in node), or in msw (or whatever over "interceptor" layer they prefer).

Mirage already does too much which is one of the reasons it's hard to maintain, so I'd potentially even be in favor of something more drastic for a major version bump where we just have `server.handle(request)` and don't worry about pretender or msw or express at all, just leave those as separate concerns other libraries can solve. As you all have said the main problem Mirage solves is the data/domain modeling side. The serializers could also be slimmed down / moved to separate libraries where they'd be easier to maintain so Mirage core could tighten up.

**Comment by andrewluetgers** [comment]
From outside looking in, just beginning to use this project, I feel that last paragraph is 100% the right way to go.

**Comment by kettanaito** [comment]
Perhaps it makes more sense to suggest isolating Mirage's modeling logic to be agnostic of any kind of underlying mocking. That way people could model their data with Mirage and use any API mocking library they want/need. 

@andrewluetgers, what you're describing sounds a lot like what we're doing in [@mswjs/data](https://github.com/mswjs/data). Have you perchance tried that library? It allows you to model your data, and it's a standalone tool, so you can use it with anything. It's not as mature as Mirage's data modeling capabilities but we're actively working on it to make it better.

**Comment by andrewluetgers** [comment]
@kettanaito thanks for the suggestion @msw/data looks great. Both of these projects are great and I want to suggest that we consider the use these modeling libraries not just for testing but for writing the applications themselves. This is what I have done in my own project, the schema is used both to define the mock api for testing and for producing a hooks based client interface for theses endpoints using react query.

**Comment by cah-briangantzler** [comment]
Sorry should have been clearer, what was still lacking that I need to complete with those PRs to get them over the finish line. So I dont have to rediscover it all over again, was there something specific that still needed done. 

They are pretty small, if you dont remember exactly then Ill rebase them and see if I see something missing

**Comment by samselikoff** [comment]
There was more to them but unfortunately I also don't remember all the details right now. Unfortunately it's just not a priority for us at the moment :(

**Comment by cah-briangantzler** [comment]
Currently working on this but having a problem with the environment that I hope someone can help with as I am fairly unfamiliar with some of it (jest tests specifically)

I use a MAC and Intellij as my editor. On the master branch all tests work. On my branch some tests fail when doing `yarn test` but for the same reason "self is not defined". In intellij when I run a specific test suite it fails for the same reason. When I then run the test suite under debug. It works. When I then run it not under debug it works most of the time. Occasionally I find a suite that will only run under debug.

The self not defined is always in the file `pretender.js`.  I found this issue that may be related. Sounds like the same issue, just not pretender but a different package, but not sure https://github.com/katspaugh/wavesurfer.js/issues/2350.

I think I almost have this completed, but not being able to reliably run the tests has slowed it down some.  Any one have any ideas?


![image](https://user-images.githubusercontent.com/1529286/161626859-2569beb1-43d2-49a9-b26c-22086db171d1.png)

**Comment by cah-briangantzler** [comment]
Looking at it more I think I might have been reading the output wrong, here is more of the output

I had thought that `resource-shorthand-test` had failed, but I think that error output was actually pointing to the red fail of `association-create-test`. I also notice that its failing under the `nodeEnvironmentConsumingEsm` config. As I look though all the failing tests, they are all under the `nodeEnvironmentConsumingEsm`. I believe that refers to this config. https://github.com/miragejs/miragejs/blob/master/jest.config.js#L27

Dont know jest enough to do anything with that information though 

I only changed code, not tests or config. So not sure why it works on master branch and not on mine. I think the reason the test pass when running under debug is that Intellij is either using some default config, or one of the ones defined but not the `nodeEnvironmentConsumingEsm` one perhaps

![image](https://user-images.githubusercontent.com/1529286/161628798-2aaf298b-82e7-4c83-b894-10150abd4a1a.png)

**Comment by cah-briangantzler** [comment]
Ok, Im guessing its because pretender doesnt work under node. Now to figure out how the test pass on master branch and how pretender ISNT part of the tests when node is running, apparently I moved it to somewhere where it is now part of the node versions of the tests.

**Comment by IanVS** [comment]
@cah-briangantzler those look like side-effect imports, since they are not actually returning any values.  The order of side-effect imports can indeed matter, since they might be modifying something on the global scope.

**Comment by samselikoff** [comment]
Yep this is what it's about, it prevents `import { x } from "miragejs"` from breaking when running in node (because importing Pretender in node throws an error).

**Comment by cah-briangantzler** [comment]
Update:

I have completely moved pretender specific code into an interceptor module. (Later will be extracted to a separate NPM package).

When Mirage server starts, if the interceptor class is not part of the options, the pretender one is used for a non-breaking change. (Will be a breaking change when the interceptor is completely removed from miragejs).

All tests pass.

Created an MSW interceptor (which polyfills msw syntax to pretender syntax so that all routes as currently defined still work). Converted my production app to use the MSW interceptor. Most tests in my app still pass, working through the few that dont. As expected there are differences between how pretender tracks and calls handlers and msw.

All in all, pretty close. Hope to have this completed soon.

**Comment by jneurock** [comment]
> Update:
> 
> I have completely moved pretender specific code into an interceptor module. (Later will be extracted to a separate NPM package).
> 
> When Mirage server starts, if the interceptor class is not part of the options, the pretender one is used for a non-breaking change. (Will be a breaking change when the interceptor is completely removed from miragejs).
> 
> All tests pass.
> 
> Created an MSW interceptor (which polyfills msw syntax to pretender syntax so that all routes as currently defined still work). Converted my production app to use the MSW interceptor. Most tests in my app still pass, working through the few that dont. As expected there are differences between how pretender tracks and calls handlers and msw.
> 
> All in all, pretty close. Hope to have this completed soon.

Nice! What an update.

**Comment by andrewluetgers** [comment]
@cah-briangantzler awesome update can't wait to try it out!

**Comment by cah-briangantzler** [comment]
I have a few tests still randomly failing. It appears to be an async problem somewhere as the tests are proceeding without waiting for the screen to be re-rendered. Since settled() is suppose to wait on promises and all fetches are promises, I would think it wouldnt matter if it was pretender or MSW. It is also random in that re-runing the tests will sometime succeed, sometimes not.

Im also sure that I am not using all possible variations of Mirage, so there may be a few rough edges yet.

I have not submitted a PR to mirage yet in case there are more changes needed and didnt want to go through a series of PRs. 

Steps to try this out in your own project (as a beta, if you want to try bleeding edge) will be posted shortly. I forked MirageJS to place where there is a permission issue. Will post as soon as Im able to move that fork.

**Comment by cah-briangantzler** [comment]
Steps to try using MSW with MirageJS (beta)

# Update your package.json miragejs to use the patched lib
```
    "miragejs": "bgantzler/miragejs#route-handlers"
```
or clone the repo `bgantzler/miragejs`. switch to branch `route-handlers` and `yarn link` (`yarn link miragejs` in your app).  If you are using `ember-cli-mirage` I found you must do the clone/link.

All your tests should still run. MirageJS will be using Pretender still

# Tell your MirageJS Server to use MSW

Add to your `package.json` the `mirage-msw` interceptor (not yet published to NPM)
```
    "mirage-msw": "bgantzler/mirage-msw",
```

Update where you define your mirage server to include a new param `interceptor`.
```js
import { Server } from 'miragejs';
import MswInterceptor from './msw-config';

new Server({
   // your config options
   interceptor: new MswInterceptor(),
});

```


Known Issues: (at least in my tests)
- Some of the tests fail to wait on the promise to succeed. The same test re-ran may work. This could be my tests, my test framework (quint) or something else. They run successfully every time under pretender.
- The logging option can only be set at server creation using the the server option `logging` passed to the Server constructor. Pretender allows you to turn this on and off after the server has been created. MSW does not until this PR https://github.com/mswjs/msw/pull/424 is landed, then will update this library to use the predicate.

**Comment by cah-briangantzler** [comment]
@andrewluetgers Would you be able to experiment with this and see if it fits the bill?

**Comment by andrewluetgers** [comment]
@cah-briangantzler will give it a go tonight

**Comment by cah-briangantzler** [comment]
I updated the example above. I had 
```js
import MswInterceptor from `./config-msw';
```
should have been
```js
import MswInterceptor from 'mirage-msw';
```

Sorry, was when i was testing locally

**Comment by cah-briangantzler** [comment]
Created a new issue https://github.com/miragejs/miragejs/issues/1037 to provide a checklist for progress towards this goal.

**Comment by IanVS** [comment]
This is a very old issue, but I wanted to let anyone subscribed here know that experimental support for MSW has been added.  The new [mirage-msw](https://github.com/miragejs/mirage-msw/tree/main) package is _mostly_ a drop-in replacement for the default pretender interceptor: 

```ts
import MSWInterceptor from 'mirage-msw';
import { createServer } from 'miragejs';

await createServer({
  interceptor: new MSWInterceptor(),
  // ... rest of your config
});
```

Note that `createServer` is now async and should be awaited (because msw is async).  You'll also need to install `msw` version 2 or higher, and run [`msw init`](https://mswjs.io/docs/integrations/browser#copy-the-worker-script) to configure it in your project.  If you decide to try it out and have any questions, feel free to post here or open an issue in the mirage-msw repo. 

I'm going to close this issue and we can track remaining work to be done in https://github.com/miragejs/miragejs/issues/1037.

**Comment by cah-brian-gantzler** [comment]
Mirage, for backward compatibility, still includes pretender, and if you dont state an interceptor, creates a pretender one.

As a breaking change, there should be a PR that moves pretender from a dependency to a peer-dependency and throws an error if an intercepter is not provided.  Since your in new versions, breaking change area on mirage.js, may want to consider that as well.

Also, hindsight 20/20 (sorry for not thinking of it sooner), for the async create server, we could have left it sync, and just moved the msw start (since it was the last line) out and just required msw people to call start (async) after create server. That may have given them more control. Unless there is a reason that it needed to be called in the create server.

**Comment by IanVS** [comment]
>for the async create server, we could have left it sync, and just moved the msw start (since it was the last line) out and just required msw people to call start (async) after create server.

I was thinking the same thing, and may still make that change, since as you say, we're in breaking-change territory.

**Comment by kettanaito** [comment]
Thank you for your work on this, @IanVS! 👏
